backend/vkt/src/main/java/fi/oph/vkt/scheduled/CancelPassiveEnrollmentsExpectingPayment.java

```diff
@@ -0,0 +1,60 @@
+package fi.oph.vkt.scheduled;
+
+import fi.oph.vkt.model.Enrollment;
+import fi.oph.vkt.model.type.EnrollmentStatus;
+import fi.oph.vkt.repository.EnrollmentRepository;
+import fi.oph.vkt.util.SchedulingUtil;
+import jakarta.annotation.Resource;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import net.javacrumbs.shedlock.spring.annotation.SchedulerLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class CancelPassiveEnrollmentsExpectingPayment {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CancelPassiveEnrollmentsExpectingPayment.class);
+
+  private static final String INITIAL_DELAY = "PT5S";
+
+  private static final String FIXED_DELAY = "PT30M";
+
+  private static final String LOCK_AT_LEAST = "PT1S";
+
+  private static final String LOCK_AT_MOST = "PT1M";
+
+  private static final Duration PAYMENT_TIME = Duration.of(3, ChronoUnit.HOURS);
+
+  @Resource
+  private final EnrollmentRepository enrollmentRepository;
+
+  @Scheduled(initialDelayString = INITIAL_DELAY, fixedDelayString = FIXED_DELAY)
+  @SchedulerLock(
+    name = "cancelPassiveEnrollmentsExpectingPayment",
+    lockAtLeastFor = LOCK_AT_LEAST,
+    lockAtMostFor = LOCK_AT_MOST
+  )
+  public void action() {
+    SchedulingUtil.runWithScheduledUser(() -> {
+      LOG.debug("cancelPassiveEnrollmentsExpectingPayment");
+
+      final List<Enrollment> enrollmentsToCancel = enrollmentRepository
+        .findAllByStatus(EnrollmentStatus.EXPECTING_PAYMENT_UNFINISHED_ENROLLMENT)
+        .stream()
+        .filter(e -> e.getModifiedAt().plus(PAYMENT_TIME).isBefore(LocalDateTime.now()))
+        .toList();
+
+      enrollmentsToCancel.forEach(e -> {
+        e.setStatus(EnrollmentStatus.CANCELED_UNFINISHED_ENROLLMENT);
+        enrollmentRepository.saveAndFlush(e);
+      });
```

Comment:
Pitäisköhän tää pätkä siirtää serviceen ja laittaa sille `@Transactional(isolation = Isolation.SERIALIZABLE)`. Teoriassa kai voisi käydä niin, että joku ehtisi tuohon väliin päivittämään enrollmenttia ja tämä pätkä ylikirjottaisi muutokset.

Reply:
Ei oo tullut tustuttua tuohon Isolation:in ominaisuuksiin / vaihtoehtoihin. Mutta kuulostaa siltä että toi olis hyvä myös muissa servicen metodeissa joissa päivitetään ilmoittautumisen tietoja? Virkailijalla on ainakin muutama eri muokkausrajapinta.

Reply:
Toimiihan tä isolation tässä myös nyt oikein jos metodi itsessään ottaa jo vastaan kannasta haetun enrollmentin?

Reply:
Tuota `Isolation.SERIALIZABLE` pitää käyttää varoen, koska se nakkaa virheen, jos joku muu on ehtiny muuttamaan tietoja. Tässä kohtaa se ei varmaan haittaa kun melko epätodennäköinen tilanne ja tuo ajetaan kuitenkin uudestaan.

En oo ihan varma mitä Spring tekee tuossa tapauksessa, jos kannasta on jo haettu jotain. 🤔 Tai siis että osaako se alottaa uuden transaktion eri isolationilla. Voisi ehkä kiertää sillä, että antais metodille enrollmentin id:n.